Add Shelly Dimmer automatic calibration

olegtarasov opened the pull request:
# What does this implement/fix?

Many dimmable light bulbs have non-linear characteristic. Original Shelly firmware
supports automatic calibration that helps make this curve more linear resulting in smoother control across the whole
brightness spectrum. An attempt was made to replicate this process in Shelly Dimmer component for ESPHome.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes https://github.com/esphome/feature-requests/issues/1711
- fixes https://github.com/esphome/issues/issues/4400

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#4443

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
light:
  - platform: shelly_dimmer
    id: dimmer
    output_id: shelly

button:
  - platform: template
    id: calibrate_button
    name: "Calibrate"
    entity_category: config
    on_press:
      then:
        - lambda: |-
            id(shelly)->start_calibration();
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

olegtarasov committed:
Add an ability to calibrate Shelly Dimmer 2

olegtarasov committed:
Merge branch 'dev' into shelly-dimmer-calibration

olegtarasov committed:
Some fixes, added an ability to clear calibration

olegtarasov committed:
Trying to keep linteer happy

olegtarasov committed:
Formatting

rnauber commented:
Hi @olegtarasov ,
thank you for your patch! I tried it on 4 non-dimmable LED and it worked flawlessly. Although I do not notice any visual difference, it seems to do its magic. There is only one thing: The calibration take quite a long time, so to not surprise anyone by that, I suggest putting a short description of what to expect from the calibration procedure in the documentation.

Thanks again for your great work, which I would love to see merged!

Best,
Richard


```
[09:18:40][D][sensor:093]: 'livingroom Power': Sending state 10.82790 W with 1 decimals of accuracy
[09:18:40][D][sensor:093]: 'livingroom Voltage': Sending state 231.55792 V with 1 decimals of accuracy
[09:18:40][D][sensor:093]: 'livingroom Current': Sending state 0.04676 A with 2 decimals of accuracy
[09:18:40][D][shelly_dimmer:597]: Calibration step 20, measurement 3
[09:18:40][D][shelly_dimmer:617]: Mean power at step 20: 10.992352
[09:18:40][D][shelly_dimmer:661]: Saved calibration to flash
[09:18:40][D][shelly_dimmer:652]: Finished calibration. Values:
[09:18:40][D][shelly_dimmer:654]: 1.000000
[09:18:40][D][shelly_dimmer:654]: 0.967905
[09:18:40][D][shelly_dimmer:654]: 0.940313
[09:18:40][D][shelly_dimmer:654]: 0.909263
[09:18:40][D][shelly_dimmer:654]: 0.874966
[09:18:40][D][shelly_dimmer:654]: 0.847662
[09:18:40][D][shelly_dimmer:654]: 0.821174
[09:18:40][D][shelly_dimmer:654]: 0.790580
[09:18:40][D][shelly_dimmer:654]: 0.754041
[09:18:40][D][shelly_dimmer:654]: 0.724630
[09:18:40][D][shelly_dimmer:654]: 0.684263
[09:18:40][D][shelly_dimmer:654]: 0.645947
[09:18:40][D][shelly_dimmer:654]: 0.612385
[09:18:40][D][shelly_dimmer:654]: 0.577788
[09:18:40][D][shelly_dimmer:654]: 0.501888
[09:18:40][D][shelly_dimmer:654]: 0.388008
[09:18:40][D][shelly_dimmer:654]: 0.256971
[09:18:40][D][shelly_dimmer:654]: 0.132151
[09:18:40][D][shelly_dimmer:654]: 0.044587
[09:18:40][D][shelly_dimmer:654]: 0.000000
....
: Remapped 0.234152 to 0.190859 (min 0.132151, max 0.256971, min_out 0.150000, max_out 0.200000)
[09:21:32][D][shelly_dimmer:188]: Remapped 0.232745 to 0.190295 (min 0.132151, max 0.256971, min_out 0.150000, max_out 0.200000)
[09:21:32][D][shelly_dimmer:188]: Remapped 0.232487 to 0.190192 (min 0.132151, max 0.256971, min_out 0.150000, max_out 0.200000)
[09:21:32][D][shelly_dimmer:188]: Remapped 0.231802 to 0.189918 (min 0.132151, max 0.256971, min_out 0.150000, max_out 0.200000)
[09:21:32][D][shelly_dimmer:200]: Brightness update: 294 (raw: 0.189918)
[09:21:32][D][shelly_dimmer:317]: Sending command: 0x01 (2 bytes) payload 0x2601
[09:21:32][D][shelly_dimmer:329]: Command sent, waiting for reply
[09:21:32][D][shelly_dimmer:439]: Got frame: 0x01
[09:21:32][D][shelly_dimmer:188]: Remapped 0.231571 to 0.189825 (min 0.132151, max 0.256971, min_out 0.150000, max_out 0.200000)
[09:21:32][D][shelly_dimmer:188]: Remapped 0.231393 to 0.189754 (min 0.132151, max 0.256971, min_out 0.150000, max_out 0.200000)
[09:21:32][D][shelly_dimmer:188]: Remapped 0.231380 to 0.189749 (min 0.132151, max 0.256971, min_out 0.150000, max_out 0.200000)
[09:21:32][D][shelly_dimmer:188]: Remapped 0.231373 to 0.189746 (min 0.132151, max 0.256971, min_out 0.150000, max_out 0.200000)
[09:21:32][D][sensor:093]: 'livingroom Uptime': Sending state 32336.51367 s with 0 decimals of accuracy




````

olegtarasov commented:
Hi @rnauber! Thank you for the review! I put a short description of calibration process near the end of documentation page for Shelly Dimmer. Here is a preview from documentation PR: https://deploy-preview-4443--esphome.netlify.app/components/light/shelly_dimmer.html. Do you think I should add further details, or is this sufficient?

olegtarasov committed:
Merge branch 'dev' into shelly-dimmer-calibration

rnauber commented:
Thanks @olegtarasov ! 

I have two questions:
1)  Would it be possible to override the update_interval automatically with a sensible default (e.g. 1s) while doing the calibration? That would allow to keep a sensible update_interval (e.g. 30s) for normal operation, while still having a fast calibration.

2) Hard coding the auto generated id "shelly_dimmer_shellydimmer_id" seems a little brittle. Can you you please modify the documentation to include an explicitly name:

```
light:
  - platform: shelly_dimmer
    output_id: shelly_dimmer_shellydimmer_id
```
Best,
Richard

olegtarasov committed:
Merge branch 'dev' into shelly-dimmer-calibration

olegtarasov committed:
Update interval is changed automatically for the duration of calibration

olegtarasov commented:
Hi @rnauber! Thanks a lot for pointing to `output_id`, I completely missed the fact that you can set in config. I've changed the documentation PR to reflect this. Also, it was actually really easy to manipulate `update_interval`, so now it's set to 1s automatically and reverted to the previous value after calibration is complete.

rnauber commented:
I tested your latest patch, @olegtarasov  and it works great! And the usability improved a lot for me, as it prevents me  from misconfiguring it now :grin: I think it is ready to get merged! Thanks again for your work!
Best.
Richard

olegtarasov committed:
Merge branch 'dev' into shelly-dimmer-calibration

rnauber reviewed (approved):
Tested and works! LGTM

olegtarasov committed:
Merge branch 'dev' into shelly-dimmer-calibration

olegtarasov committed:
Merge branch 'dev' into shelly-dimmer-calibration

olegtarasov committed:
Merge branch 'dev' into shelly-dimmer-calibration

olegtarasov committed:
Merge branch 'dev' into shelly-dimmer-calibration

olegtarasov committed:
Merge branch 'dev' into shelly-dimmer-calibration

olegtarasov committed:
Merge branch 'dev' into shelly-dimmer-calibration

olegtarasov committed:
Merge branch 'dev' into shelly-dimmer-calibration

olegtarasov committed:
Merge branch 'dev' into shelly-dimmer-calibration

olegtarasov committed:
Merge branch 'dev' into shelly-dimmer-calibration

olegtarasov committed:
Merge branch 'dev' into shelly-dimmer-calibration

olegtarasov committed:
Merge branch 'dev' into shelly-dimmer-calibration

olegtarasov committed:
Merge branch 'dev' into shelly-dimmer-calibration

olegtarasov committed:
Merge branch 'dev' into shelly-dimmer-calibration

edge90 commented on esphome/components/shelly_dimmer/shelly_dimmer.cpp:
```diff
@@ -148,6 +168,32 @@ void ShellyDimmer::write_state(light::LightState *state) {
   state->current_values_as_brightness(&brightness);
 
+  // If we are in a process of calibration, don't mess with brightness.
+  // Also check whether we have calibration data and wheteher edge values were requested.
```
plz fix

edge90 commented on esphome/components/shelly_dimmer/shelly_dimmer.cpp:
```diff
@@ -523,6 +569,125 @@ void ShellyDimmer::reset_dfu_boot_() {
+  this->calibration_measurements_.fill(0);
+  this->calibration_measurement_cnt_ = 0;
+  this->calibration_step_ = -20;
+  this->calibrating_ = true;
```
You don't need `this->`.

edge90 reviewed (commented):
Looks almost good.

Note: Havn't tested it on a real device.

edge90 commented on esphome/components/shelly_dimmer/shelly_dimmer.cpp:
```diff
@@ -148,6 +168,32 @@ void ShellyDimmer::write_state(light::LightState *state) {
+    } else {
+      float min = this->calibration_data_[pos];
+      float max = this->calibration_data_[pos - 1];
+      float min_out = 1 - (float) pos * CALIBRATION_STEP;
```
Don't use c-style casts.

rnauber commented:
Hi everyone, 
I just want to report that this works really well on a real device. I would love to see it merged.

Best,
Richard

olegtarasov committed:
Merge branch 'dev' into shelly-dimmer-calibration
